syntax = "proto3";

package ProtobufClient;

// ClientMessage is a wrapper message used to describe the supplied client message
// and pass metadata about the client.
message ClientMessage {
  oneof body {
    ClientIdentificationMessage identification = 1 [deprecated=true];
    ClientSubscribeMessage subscribe = 2;
    ClientUnsubscribeMessage unsubscribe = 3;
    WebAuthenticationMessage webAuthentication = 4;
    APIAuthenticationMessage apiAuthentication = 5;
    ClientSessionMessage clientSession = 6;
  }
}

// ClientIdentificationMessage is the first message sent is from the client to
// the server, identifying itself and the subscriptions it desires.
message ClientIdentificationMessage {
  string useragent = 1;
  string revision = 2;
  string integration = 3;
  string locale = 4;
  repeated string subscriptions = 5 [deprecated=true];
  repeated ClientSubscription clientSubscriptions = 6; // should be used for subscriptions
}

message WebAuthenticationMessage {
  ClientIdentificationMessage identification = 1;
  string token = 2;
  string nonce = 3;
  repeated string access_list = 4;
}

message WebAuthenticationResult {
  Status status = 1;
  enum Status {
    UNKNOWN = 0;
    AUTHENTICATED = 1;
    INVALID_SESSION = 2;
    MFA_REQUIRED = 3;
  }
}

message TradeSessionAuth {
  string api_key = 1;
  string api_secret = 2;
  string customer_id = 3;
  string key_passphrase = 4;
}

message TradeSubscription {
  string marketId = 1;

  // optional: use if supplying different exchange keys
  TradeSessionAuth auth = 2;
}

message StreamSubscription {
  string resource = 1;
}

message UserPushSubscription {
  message TriggeredHandlers {}
  message Achievements {}
  message BalanceUpdate {}

  oneof body {
    // receive websocket pushes from trigger
    TriggeredHandlers triggered_handlers = 1;
    Achievements achievements = 2;
    BalanceUpdate balance_update = 3;
  }
}

// DEPRECATED the UserPushSubscription now wraps trigger-related subscriptions.
message TriggerSubscription {}

message ClientSubscription {
  oneof body {
    StreamSubscription stream_subscription = 1;
    TradeSubscription trade_subscription = 2;
    TriggerSubscription trigger_subscription = 3 [deprecated=true];
    UserPushSubscription user_push_subscription = 4;
  }
}

message APIAuthenticationMessage {
  string token = 1;
  string nonce = 2;
  string api_key = 3;
  Source source = 4;
  string version = 5;
  repeated string subscriptions = 6 [deprecated=true]; // kept for compability
  repeated ClientSubscription client_subscriptions = 7; // should be used for subscriptions

  enum Source {
    UNKNOWN = 0;
    GOLANG_SDK = 1;
    JAVASCRIPT_SDK = 2;
    NODE_SDK = 3;
    RUST_SDK = 5;
    CW_WEB = 4;
  }
}

message ClientSessionMessage {
  message Session {
    string userId = 1;
    int64 expires = 2;
    string token = 3;
    string mfaToken = 4;
  }

  message AnonymousTradingSession {
    string exchange = 1;
    string token = 2;
    int64 expiration = 3;
  }

  message AnonymousUserSession {
    string token = 1; // expected to be a user.AnonToken
  }

  oneof SessionConfig {
    Session session = 1;
    AnonymousTradingSession anonymousTradingSession = 2;
    AnonymousUserSession anonymousUserSession = 4;
  }

  ClientIdentificationMessage identification = 3;
}

// SubscribeMessage informs the Stream Hub to subscribe the current web socket to
// the supplied channel.
message ClientSubscribeMessage {
  repeated string subscriptionKeys = 1 [ deprecated = true ]; // kept for compability
  repeated ClientSubscription subscriptions = 2; // list of subscriptions
}

// UnsubscribeMessage informs the Stream Hub to unsubscribe the current web socket
// from the supplied channel.
message ClientUnsubscribeMessage {
  repeated string subscriptionKeys = 1 [ deprecated = true ]; // kept for compability
  repeated ClientSubscription subscriptions = 2; // list of subscriptions
}
